Realm/Realm/DatabaseTypes/KeyPathCollection.cs

```diff
@@ -0,0 +1,114 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Realms;
+
+//TODO Add docs for this and the following types
+public class KeyPathsCollection : IEnumerable<KeyPath>
+{
+    private IEnumerable<KeyPath> _collection;
+
+    private static readonly KeyPathsCollection _shallow = new KeyPathsCollection(KeyPathsCollectionType.Shallow);
+    private static readonly KeyPathsCollection _default = new KeyPathsCollection(KeyPathsCollectionType.Default);
+
+    internal KeyPathsCollectionType Type { get; set; }
+
+    private KeyPathsCollection(KeyPathsCollectionType type, IEnumerable<KeyPath>? collection = null)
+    {
+        Debug.Assert(type == KeyPathsCollectionType.Full == (collection?.Any() == true), "If collection isn't empty, then the type must be Full");
+
+        Type = type;
+        _collection = collection ?? Enumerable.Empty<KeyPath>();
+
+        VerifyKeyPaths();
+    }
+
+    internal IEnumerable<string> GetStrings() => _collection.Select(x => x.Path);
+
+    internal void VerifyKeyPaths()
+    {
+        foreach (var item in _collection)
+        {
+            if (string.IsNullOrWhiteSpace(item?.Path))
+            {
+                throw new ArgumentException("A key path cannot be null, empty, or consisting only of white spaces");
+            }
+        }
+    }
+
+    public static KeyPathsCollection Of(params KeyPath[] paths)
+    {
+        if (paths.Length == 0)
+        {
+            return new KeyPathsCollection(KeyPathsCollectionType.Shallow);
+        }
+
+        return new KeyPathsCollection(KeyPathsCollectionType.Full, paths);
+    }
+
+    public static KeyPathsCollection Shallow => _shallow;
+
+    public static KeyPathsCollection Default => _default;
+
+    public static implicit operator KeyPathsCollection(List<string> paths) =>
+        new(KeyPathsCollectionType.Full, paths.Select(path => (KeyPath)path));
+
+    public static implicit operator KeyPathsCollection(List<KeyPath> paths) => new(KeyPathsCollectionType.Full, paths);
+
+    public static implicit operator KeyPathsCollection(string[] paths) =>
+        new(KeyPathsCollectionType.Full, paths.Select(path => (KeyPath)path));
+
+    public static implicit operator KeyPathsCollection(KeyPath[] paths) => new(KeyPathsCollectionType.Full, paths);
+
+    /// <inheritdoc/>
+    public IEnumerator<KeyPath> GetEnumerator()
+    {
+        return _collection.GetEnumerator();
+    }
+
+    /// <inheritdoc/>
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+
+public class KeyPath
+{
+    internal string Path { get; private set; }
+
+    private KeyPath(string path)
+    {
+        Path = path;
+    }
+
+    public static implicit operator KeyPath(string s) => new(s);
```

Comment:
Should we add an implicit string(KeyPath) operator?

Reply:
I didn't add it because I didn't think it would be as useful as the opposite

Reply:
I agree - it's not clear to me that it's very useful either, just figured it's kind of weird that you construct an object but then can't get the info out of it. Again, I don't see any practical application to it, just satisfying my OCD 😅